Update webrtc-stats according to the latest spec

jyavenard opened the pull request:
#### 066a9fc870b06ca5eed91af3c7ad170e253242af
<pre>
Update webrtc-stats according to the latest spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=262881">https://bugs.webkit.org/show_bug.cgi?id=262881</a>
rdar://116670935

Reviewed by Youenn Fablet.

Update WebRTC stats according to the latest spec editor&apos;s draft (as of August 4th 2023)
and re-sync with WPT.
<a href="https://www.w3.org/TR/webrtc-stats">https://www.w3.org/TR/webrtc-stats</a>

Some tests are still failing, because WPT aren&apos;t testing the actual specs.
Stats that requires the user to confirm the exchange will be implemented
in a follow-up updates.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/README.md: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/getStats-remote-candidate-address-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/getStats-remote-candidate-address.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/hardware-capability-stats.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/hardware-capability-stats.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/idlharness.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/idlharness.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/idlharness.window.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/outbound-rtp.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/outbound-rtp.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/rtp-stats-creation-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/rtp-stats-creation.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/supported-stats.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/supported-stats.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-mandatory-getStats.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-transport-stats.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCRtpSender-getStats.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/protocol/dtls-setup.https-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/tests-options.json:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.exp:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/mediastream/RTCIceCandidate.h:
* Source/WebCore/Modules/mediastream/RTCIceRole.h: Added.
* Source/WebCore/Modules/mediastream/RTCIceRole.idl: Added.
* Source/WebCore/Modules/mediastream/RTCIceServerTransportProtocol.h: Added.
* Source/WebCore/Modules/mediastream/RTCIceServerTransportProtocol.idl: Added.
* Source/WebCore/Modules/mediastream/RTCStatsReport.h:
(WebCore::RTCStatsReport::RemoteOutboundRtpStreamStats::RemoteOutboundRtpStreamStats):
(WebCore::RTCStatsReport::AudioPlayoutStats::AudioPlayoutStats):
* Source/WebCore/Modules/mediastream/RTCStatsReport.idl:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerStatsCollector.cpp:
(WebCore::fillReceivedRTPStreamStats):
(WebCore::iceCandidateType):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCStatsCollector.cpp:
(WebCore::fillRtpStreamStats):
(WebCore::fillReceivedRtpStreamStats):
(WebCore::fillInboundRtpStreamStats):
(WebCore::fillRemoteInboundRtpStreamStats):
(WebCore::fillSentRtpStreamStats):
(WebCore::qualityLimitationReason):
(WebCore::fillOutboundRtpStreamStats):
(WebCore::fillRemoteOutboundRtpStreamStats):
(WebCore::iceCandidatePairState):
(WebCore::fillRTCIceCandidatePairStats):
(WebCore::iceCandidateState):
(WebCore::fillRTCIceCandidateStats):
(WebCore::fillRTCCodecStats):
(WebCore::iceRole):
(WebCore::dtlsTransportState):
(WebCore::iceTransportState):
(WebCore::dtlsRole):
(WebCore::fillRTCTransportStats):
(WebCore::fillRTCAudioSourceStats):
(WebCore::fillRTCAudioPlayoutStats):
(WebCore::initializeRTCStatsReportBackingMap):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/269353@main">https://commits.webkit.org/269353@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd3be11d1ad29126064074bbd6a9fbe426430143

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22320 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22518 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23395 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24220 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20657 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22570 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26780 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22849 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21665 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22556 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22252 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19372 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25075 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19302 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20227 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26479 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20319 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20458 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24336 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20989 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17781 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20446 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20036 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5314 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24677 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20984 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on LayoutTests/platform/mac-wk2/TestExpectations:
```diff
@@ -2084,6 +2084,9 @@ webkit.org/b/261215 webrtc/vp8-then-h264-gpu-process-crash.html [ Pass Timeout F
 webkit.org/b/261226 [ Sonoma+ ] fast/images/mac/play-pause-individual-animation-context-menu-items.html [ Pass ] # backing functionality only in Sonoma+
 
+# The test takes an incredibly long time to run and typically timesout when run in stress-test, so skip it for now.
+webkit.org/b/263044 imported/w3c/web-platform-tests/webrtc-stats/supported-stats.https.html [ Skip ]
```
Let's keep it for release, maybe PASS/FAIL is fine.

youennf commented on LayoutTests/TestExpectations:
```diff
@@ -2299,8 +2299,8 @@ imported/w3c/web-platform-tests/webrtc/protocol/split.https.html [ Pass Failure
 imported/w3c/web-platform-tests/webrtc/RTCRtpParameters-encodings.html [ Pass Failure ]
-
 imported/w3c/web-platform-tests/mst-content-hint/RTCRtpSendParameters-degradationEffect.html [ Failure ]
+imported/w3c/web-platform-tests/webrtc-stats/supported-stats.https.html [ Slow ]
```
Not needed I think given test-options.json

youennf replied:
Please check whether this is actually needed by running rwt. If not, I suggest removing it.

youennf commented on Source/WebCore/Modules/mediastream/RTCIceCandidate.idl:
```diff
@@ -51,6 +51,7 @@
     readonly attribute DOMString? relatedAddress;
     readonly attribute unsigned short? relatedPort;
     readonly attribute DOMString? usernameFragment;
+    readonly attribute RTCIceServerTransportProtocol? relayProtocol;
```
This is new, is it needed in this patch? Where is it covered?

jyavenard replied:
it was one reason some of the existing tests were failing. I've removed it from now and will cover it in another PR

youennf commented on Source/WebCore/Modules/mediastream/RTCStatsReport.idl:
```diff
@@ -120,6 +118,15 @@ dictionary RTCInboundRtpStreamStats : RTCReceivedRtpStreamStats {
     unsigned long framesReceived;
+    DOMString decoderImplementation;
+    DOMString playoutId;
+    boolean powerEfficientDecoder;
```
We do not really like this one for now, let's not add it in this PR.

jyavenard replied:
it's in the spec, the IDL is there, however it's not populated, so I think it should say so that we pass the IDL compliance test

jyavenard replied:
it's in the spec, the IDL is there, however it's not populated, so I think it should say so that we pass the IDL compliance test

youennf commented on Source/WebCore/Modules/mediastream/RTCStatsReport.idl:
```diff
@@ -120,6 +118,15 @@ dictionary RTCInboundRtpStreamStats : RTCReceivedRtpStreamStats {
     double totalAudioEnergy;
     double totalSamplesDuration;
     unsigned long framesReceived;
+    DOMString decoderImplementation;
```
We do not really like this one for now, let's not add it in this PR.

jyavenard replied:
same

youennf commented on Source/WebCore/Modules/mediastream/RTCStatsReport.idl:
```diff
@@ -138,181 +144,208 @@ dictionary RTCSentRtpStreamStats : RTCRtpStreamStats {
-    unsigned long sliCount;
-    // DOMString encoderImplementation;
+    DOMString encoderImplementation;
+    boolean powerEfficientEncoder;
```
Ditto for these two.

jyavenard replied:
same

youennf reviewed (commented):
Looks good to me, some comments before landing.

youennf commented on LayoutTests/platform/mac-wk2/TestExpectations:
```diff
@@ -1901,6 +1901,9 @@ webkit.org/b/261215 webrtc/vp8-then-h264-gpu-process-crash.html [ Pass Timeout F
 
 webkit.org/b/261226 [ Sonoma+ ] fast/images/mac/play-pause-individual-animation-context-menu-items.html [ Pass ] # backing functionality only in Sonoma+
 
+# The test takes an incredibly long time to run and typically timesout when run in stress-test.
```
```suggestion
# The test takes an incredibly long time to run and typically times out when run in stress-test.
```

youennf commented on Source/WebCore/Modules/mediastream/RTCIceCandidate.h:
```diff
@@ -64,7 +64,6 @@ class RTCIceCandidate final : public RefCounted<RTCIceCandidate>, public ScriptW
     String relatedAddress() const { return m_fields.relatedAddress; }
     std::optional<unsigned short> relatedPort() const { return m_fields.relatedPort; }
     String usernameFragment() const { return m_fields.usernameFragment; }
-
```
Change unrelated.

youennf commented on Source/WebCore/Modules/mediastream/RTCStatsReport.h:
```diff
@@ -145,54 +156,62 @@ class RTCStatsReport : public RefCounted<RTCStatsReport> {
-        std::optional<uint32_t> sliCount;
-        // DOMString encoderImplementation;
+        String encoderImplementation;
+        std::optional<bool> powerEfficientEncoder;
```
This increases the size of the struct while not being used.
I would remove them until we have a plan to expose them somehow.

youennf commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCStatsCollector.cpp:
```diff
@@ -62,47 +63,48 @@ static inline void fillRtpStreamStats(RTCStatsReport::RtpStreamStats& stats, con
         stats.keyFramesDecoded = *rtcStats.key_frames_decoded;
+    // Not Implemented
+    // if (rtcStats.frames_rendered.is_defined())
+    //     stats.framesRendered = *rtcStats.frames_rendered;
```
I am not sure we want to add these lines, mentioning a comment like // FIXME: Add framesRendered might be good enough.
Ditto below.

youennf commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCStatsCollector.cpp:
```diff
@@ -379,10 +459,12 @@ static inline void fillRTCIceCandidateStats(RTCStatsReport::IceCandidateStats& s
-        stats.candidateType = iceCandidateState(*rtcStats.candidate_type);
+    ASSERT(rtcStats.candidate_type.is_defined());
+    auto candidateType = iceCandidateState(*rtcStats.candidate_type);
+    ASSERT(candidateType);
```
We might want to modify iceCandidateState to return a RTCIceCandidateType instead of an optional

youennf commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCStatsCollector.cpp:
```diff
@@ -421,31 +505,103 @@ static inline void fillRTCCodecStats(RTCStatsReport::CodecStats& stats, const we
         stats.sdpFmtpLine = fromStdString(*rtcStats.sdp_fmtp_line);
 }
 
+static inline std::optional<RTCIceRole> iceRole(const std::string& state)
```
Ditto here, we might stick with returning a RTCIceRole, ditto below

youennf commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCUtils.cpp:
```diff
@@ -410,6 +410,18 @@ static inline std::optional<RTCIceCandidateType> toRTCIceCandidateType(const std
     return RTCIceCandidateType::Relay;
 }
 
+static inline std::optional<RTCIceServerTransportProtocol> toRTCIceServerTransportProtocol(const std::string& protocol)
```
Needed?

jyavenard committed:
Update webrtc-stats according to the latest spec
https://bugs.webkit.org/show_bug.cgi?id=262881
rdar://116670935

Reviewed by Youenn Fablet.

Update WebRTC stats according to the latest spec editor's draft (as of August 4th 2023)
and re-sync with WPT.
https://www.w3.org/TR/webrtc-stats

Some tests are still failing, because WPT aren't testing the actual specs.
Stats that requires the user to confirm the exchange will be implemented
in a follow-up updates.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/README.md: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/getStats-remote-candidate-address-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/getStats-remote-candidate-address.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/hardware-capability-stats.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/hardware-capability-stats.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/idlharness.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/idlharness.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/idlharness.window.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/outbound-rtp.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/outbound-rtp.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/rtp-stats-creation-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/rtp-stats-creation.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/supported-stats.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/supported-stats.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-stats/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-mandatory-getStats.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-transport-stats.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCRtpSender-getStats.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/protocol/dtls-setup.https-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/tests-options.json:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.exp:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/mediastream/RTCIceCandidate.h:
* Source/WebCore/Modules/mediastream/RTCIceRole.h: Added.
* Source/WebCore/Modules/mediastream/RTCIceRole.idl: Added.
* Source/WebCore/Modules/mediastream/RTCIceServerTransportProtocol.h: Added.
* Source/WebCore/Modules/mediastream/RTCIceServerTransportProtocol.idl: Added.
* Source/WebCore/Modules/mediastream/RTCStatsReport.h:
(WebCore::RTCStatsReport::RemoteOutboundRtpStreamStats::RemoteOutboundRtpStreamStats):
(WebCore::RTCStatsReport::AudioPlayoutStats::AudioPlayoutStats):
* Source/WebCore/Modules/mediastream/RTCStatsReport.idl:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerStatsCollector.cpp:
(WebCore::fillReceivedRTPStreamStats):
(WebCore::iceCandidateType):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCStatsCollector.cpp:
(WebCore::fillRtpStreamStats):
(WebCore::fillReceivedRtpStreamStats):
(WebCore::fillInboundRtpStreamStats):
(WebCore::fillRemoteInboundRtpStreamStats):
(WebCore::fillSentRtpStreamStats):
(WebCore::qualityLimitationReason):
(WebCore::fillOutboundRtpStreamStats):
(WebCore::fillRemoteOutboundRtpStreamStats):
(WebCore::iceCandidatePairState):
(WebCore::fillRTCIceCandidatePairStats):
(WebCore::iceCandidateState):
(WebCore::fillRTCIceCandidateStats):
(WebCore::fillRTCCodecStats):
(WebCore::iceRole):
(WebCore::dtlsTransportState):
(WebCore::iceTransportState):
(WebCore::dtlsRole):
(WebCore::fillRTCTransportStats):
(WebCore::fillRTCAudioSourceStats):
(WebCore::fillRTCAudioPlayoutStats):
(WebCore::initializeRTCStatsReportBackingMap):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/269353@main
